Add ESQL match function

carlosdelest opened the pull request:
Adds a `match()` function, behind snapshot builds.

This is similar to the work done on https://github.com/elastic/elasticsearch/pull/112590. It adds a new `FullTextFunction` for match queries.

This PR modifies the ESQL grammar to allow `match` to be a valid function identifier - as the `MATCH` operator introduced the `MATCH` token.

Some reworking needed to be done in `FullTextFunction` for adding multiple ordinal parameters support.

It has the same limitations as QSTR - `FullTextFunctions` need to be pushed down to Lucene as part of the local physical plan optimizing. We're limiting the commands that can be used before the function, so we don't allow commands that may alter the existing columns (DROP, KEEP, EVAL).

Commit:
First version

Commit:
Wire up for capabilities and function registry

Commit:
Minor text message change

Commit:
Changed hierarchy to allow for more expressions besides query

Commit:
Change name to matchstr in registry

Commit:
Added VerifierTests for matchstr

Commit:
Fix TypeResolution method

Commit:
Accept only a field as a first argument

Commit:
Add CSV tests

Commit:
Modified type errors to include source and not function name

Commit:
Don't do CSV tests for match function

Commit:
Add MatchFunctionTests

Commit:
Add generated docs

Commit:
Make PR easier to read

carlosdelest commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -155,18 +155,19 @@ public static TypeResolution isNotNullAndFoldable(Expression e, String operation
         return resolution;
     }
 
-    public static TypeResolution isNotFoldable(Expression e, String operationName, ParamOrdinal paramOrd) {
```
Removed `isNotFoldable()` method as it was unused, and created `isNotNull()`

Commit:
Minor refactorings for better javadoc and subclass structure for FullTextFunctions

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1107,86 +1108,155 @@ private void assertMatchCommand(String lineAndColumn, String command, String que
-        // Source commands
-        assertEquals("1:13: [QSTR] function cannot be used after SHOW", error("show info | where qstr(\"8.16.0\")"));
+        assertEquals("1:13: [QSTR] function cannot be used after SHOW", error("show info | where qstr(\"Anna\")"));
         assertEquals("1:17: [QSTR] function cannot be used after ROW", error("row a= \"Anna\" | where qstr(\"Anna\")"));
```
Tests for match and qstr have been refactored so they are shared where possible

Commit:
Spotless

Commit:
Check function defined in tests

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java

Commit:
Rename match operator to matchstr

Commit:
Rename matchstr function to match

Commit:
Rename matchstr function to match

Commit:
Rename match operator to matchstr

Commit:
Don't run REST compat test for match operator

Commit:
Revert "Rename match operator to matchstr"

This reverts commit 0a05d5b79881fbf8df094b3e4520fcfc6c384c31.

Commit:
Revert "Rename match operator to matchstr"

This reverts commit e23eef3f

Commit:
Adds MATCH as a valid function identifier to make it compatible with match operator

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

# Conflicts:
#	x-pack/plugin/build.gradle

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/FullTextFunction.java:
```diff
@@ -38,43 +35,61 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
 
-    protected FullTextFunction(Source source, Expression query) {
-        super(source, singletonList(query));
+    protected FullTextFunction(Source source, Expression query, Expression... nonQueryChildren) {
```
`FullTextFunction` has been modified to allow multiple children expressions

Commit:
Removed unnecessary code

Commit:
Final removal of previous renames

carlosdelest committed:
Update docs/changelog/113374.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @carlosdelest, I've created a changelog YAML for you.

ioanatia commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1079,83 +1080,155 @@ public void testMatchFilter() throws Exception {
-            "1:52: [QSTR] function cannot be used after STATS",
-            error("from test | STATS c = COUNT(emp_no) BY languages | where qstr(\"Anna\")")
+            "1:31: [" + functionName + "] function cannot be used after KEEP",
+            error("from test | keep first_name | where " + functionInvocation)
```
this should work for match `from test | keep first_name | where match(first_name, "Anna")`
the same if `match` is used in a `where` after a `drop`.

ioanatia commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1079,83 +1080,155 @@ public void testMatchFilter() throws Exception {
+            "1:27: [" + functionName + "] function cannot be used after DROP",
+            error("from test | drop emp_no | where " + functionInvocation)
         );
-        assertEquals("1:26: [QSTR] function cannot be used after EVAL", error("from test | eval z = 2 | where qstr(\"Anna\")"));
```
Do we still have a test for `EVAL` for QSTR?
For `match` we would want to allow it to be used after an `EVAL` - but we would need to explicitly check that the field it receives is a field present in the mappings similar to what we did for the `match` operator:

https://github.com/elastic/elasticsearch/blob/ade235bf0f032926ab611a17d553160bc2f254a9/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java#L622

ioanatia reviewed (commented):
I think we need an additonal check done on the verifier that the query can be pushed to lucene. We did that for the `match` operator.

Without it when the query cannot be pushed to Lucene we fail at the data node level with a 500 and that is too late in the execution - that's why it would be good to do it at the `Verifier` level:

```
 FROM test-index | WHERE qstr("aaaa") OR MV_COUNT(category) > 1
```
would return:
```
{
  "error": {
    "root_cause": [
      {
        "type": "ql_illegal_argument_exception",
        "reason": """Unsupported expression [qstr("aaaa")]""",
        "suppressed": [
          {
            "type": "exception",
            "reason": "1 further exceptions were dropped"
          },
          {
            "type": "task_cancelled_exception",
            "reason": "parent task was cancelled [cancelled on failure]"
          }
        ]
      }
    ],
    "type": "ql_illegal_argument_exception",
    "reason": """Unsupported expression [qstr("aaaa")]""",
    "suppressed": [
      {
        "type": "exception",
        "reason": "1 further exceptions were dropped"
      },
      {
        "type": "task_cancelled_exception",
        "reason": "parent task was cancelled [cancelled on failure]"
      }
    ]
  },
  "status": 500
}
```
I assume the same would happen if I test the `match` function with this branch (I did not test).

the lucene pushdown check for the operator is done in:

https://github.com/elastic/elasticsearch/blob/ade235bf0f032926ab611a17d553160bc2f254a9/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java#L636

Commit:
Match function needs to take into account field information instead of failing after commands

Commit:
Add verifier tests - remove tests for executing after commands for match()

Commit:
Add tests for using match() after commands

Commit:
Spotless

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

Commit:
Fix tests

Commit:
Check that conditions in full text functions are all pushable to Lucene

Commit:
Add LocalPhysicalPlanOptimizerTests for match function

Commit:
Remove test with non-pushed down function

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

Commit:
Change function description

Commit:
Right, Spotless

Commit:
Reword error message

Commit:
Improve condition performance

Commit:
Refactor checks for asQuery() method

Commit:
Fix docs

carlosdelest commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -63,6 +63,18 @@ public static TypeResolution isString(Expression e, String operationName, ParamO
+        }
+
+        return new TypeResolution(
+            format(null, "[{}] cannot operate on [{}] which is not an Elasticsearch field", operationName, e.sourceText())
```
Done in 0cf99be63600aefe91bdf0318c27ea4374d18627

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -661,7 +667,21 @@ private static void checkFullTextQueryFunctions(LogicalPlan plan, Set<Failure> f
                 });
-            }
+
+                if ((canPushToSource(condition, x -> false) == false) && hasFullTextFunction.get()) {
```
That's better! Done in 87609922d8863afbfc006d34d15689e6cd0566aa

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,125 @@
+    @FunctionInfo(
+        returnType = "boolean",
+        preview = true,
+        description = "Performs a match query on the specified fields. Returns true if the provided query matches the row.",
```
🤦 it's singular. Fixed in 67a9fb05d79b7eb73c02409460b7ecfeed697e3d, and docs in a976d2c1c0556722ee98fe1dae854ee99df141e1

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,125 @@
+    public Query asQuery() {
+        Object queryAsObject = query().fold();
+        if (queryAsObject instanceof BytesRef == false) {
+            throw new IllegalArgumentException("Query in MATCH function needs to be resolved to a string");
```
I don't think it's possible to hit these errors, as they are already being checked on the `resolveQueryType()` + `resolveNonQueryParamTypes()` methods. 

I've fixed the error messages and done a small refactoring to remove duplicated code in 67a9fb05d79b7eb73c02409460b7ecfeed697e3d

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -661,7 +667,21 @@ private static void checkFullTextQueryFunctions(LogicalPlan plan, Set<Failure> f
                 });
-            }
+
+                if ((canPushToSource(condition, x -> false) == false) && hasFullTextFunction.get()) {
```
I refactored this a bit so that may be why it looked wrong - it's indeed after the `condition.forEachDown()`

ioanatia replied:
I pulled the branch with the latest changes and it looks like we make this check in `condition.forEachDown`:

<img width="1147" alt="Screenshot 2024-10-01 at 16 43 06" src="https://github.com/user-attachments/assets/76d4cd76-d800-41fb-9884-0ff6eee166e3">

carlosdelest replied:
Ah, I see what you mean. Thanks for checking this out.

I've refactored the code in 41dc6d393e65a614dbd2cc81e738e841494ea7e1 to make sure we look first for a full text function in the condition, and then perform the necessary checks.

Does that look better?

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

Commit:
Refactor verifier code

fang-xing-esql reviewed (commented):
Thanks for putting these together @carlosdelest . I added some comments for the grammar and the functions themselves. 

I'll take a look at `Verifier` and related tests and add more comments in the next round.

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,120 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+
+        assertEquals(
+            "1:47: [match(y, \"Anna\")] cannot operate on [y], which is not a field from an index mapping",
+            error("from test | eval y = concat(\"a\", \"b\") | where match(y, \"Anna\"" + ")")
```
Seems like the `+` is unnecessary, same to line 1260, 1262.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +645,50 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
+        // Similar to cases present in org.elasticsearch.xpack.esql.optimizer.rules.PushDownAndCombineFilters -
+        // we can't check if it can be pushed down if we don't have information about the fields present in the query.
+        plan.forEachDown(LogicalPlan.class, lp -> {
+            if ((lp instanceof Filter || lp instanceof OrderBy || lp instanceof EsRelation || ftf.hasFieldsInformation()) == false) {
```
I found the checking to `ftf.hasFieldsInformation()` a bit confusing, I wonder if it means these restrictions apply to `QueryStringFunction` only? Does it mean `MatchFunction` can be after any processing commands? 

Some comments will be helpful, or if this applies to `QueryStringFunction` only and `hasFieldsInformation()` is only used here, perhaps we can check if `ftf` is an instance of `QueryStringFunction` explicitly here and get rid of `hasFieldsInformation()`? 

If this applies to `QueryStringFunction` only, perhaps rename the method to something like `validateQueryStringFunctionPosition`? There could be a better method name, to make it more self-explained.

carlosdelest replied:
For now it only applies to MatchFunction. As this is confusing, I changed it in fedefcea574f09e6260cbc233e8b936e0376420e to use `instanceof`.

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +645,50 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
     }
+
+    private static void checkCommandUsageAfterFullTextFunction(FullTextFunction ftf, LogicalPlan plan, Set<Failure> failures) {
+        // Similar to cases present in org.elasticsearch.xpack.esql.optimizer.rules.PushDownAndCombineFilters -
```
These comments seem not quite relevant to this method?

carlosdelest replied:
You're right - I updated it in fedefcea574f09e6260cbc233e8b936e0376420e

fang-xing-esql reviewed (commented):
Some more comments to `Verifier` and related tests are added.

ioanatia commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:
```diff
@@ -88,19 +88,6 @@ card:keyword   |host:keyword   |ip0:ip                   |ip1:ip
 eth1           |beta           |127.0.0.1                |127.0.0.2
 ;
 
-qstrWithFunctionNotPushedToLucene
```
Our plan is that this restriction will be eventually lifted. We have prototyped how executing full text functions will look on the compute engine already and it is not a trivial/small change that we can include in this PR.
We already discussed that in the initial release of the match/qstr functions we will have restrictions and error out when they cannot be pushed to Lucene. The checks that Carlos has added here were already present for the match operator, so they are not new.
The other aspect is that we will introduce scoring which will further complexify the eventual compute engine implementation.

carlosdelest replied:
I added support for having pushable / non pushable conjunctions, as well as pushable disjunctions.

Commit:
Implement Validatable for checking that field is an Elasticsearch field

Commit:
Grammar changes from PR review

Commit:
Change FullTextFunction constructor signature

Commit:
Centralize writeTo in FullTextFunction

Commit:
Rename resolveQueryType, add javadoc

Commit:
Distinguish between MatchFunction and QueryStringFunction for pushing down

Commit:
Remove hasFieldsInformation() and revert to an instanceof comparison

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +645,50 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
+    }
+
+    private static void checkConditionCanBePushedDown(FullTextFunction ftf, Expression condition, Set<Failure> failures) {
+        if (canPushToSource(condition, x -> false) == false) {
```
I see your point. I have changed the implementation to distinguish between the two in 1380c706b4e7a3ab379b35ed77168f835166a642.

> MatchFunction has field information, does it worth checking MatchFunction 's field further in canPushToSource, similar to CIDRMatch?

I've implemented the same checks as in the MATCH operator. This does not require the field to be exact, just to be of string type.

> And canPushToSource usually runs on data nodes, can leverage SearchStats there, however Verifier runs on the coordinator node, is SearchStats available on the coordinator node?

The `hasIdenticalDelegate()` is not checked in the code paths I've seen (it always return false). I'm not sure what would be the need of SearchStats?

fang-xing-esql replied:
I'm still not quite unclear about whether it is appropriate to call `canPushToSource` here, as it is usually called from the data node by local optimizer, not coordinator node. It seems like it affects `TEXT` field more. For example, do we expect this query to work?
```
curl -u elastic:password -X PUT "localhost:9200/books?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "field1": {
        "type": "text"
      },
      "field2": {
        "type": "text"
      }
    }
  }
}
'

+ curl -u elastic:password -v -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from books | where match(field1, \"a\") or field2 == \"b\""
}
'

{
  "error" : {
    "root_cause" : [
      {
        "type" : "verification_exception",
        "reason" : "Found 1 problem\nline 1:20: Invalid condition [match(field1, \"a\") or field2 == \"b\"]. Function MATCH can't be used as part of an or condition that includes [field2 == \"b\"]"
      }
    ],
    "type" : "verification_exception",
    "reason" : "Found 1 problem\nline 1:20: Invalid condition [match(field1, \"a\") or field2 == \"b\"]. Function MATCH can't be used as part of an or condition that includes [field2 == \"b\"]"
  },
  "status" : 400
}
* Connection #0 to host localhost left intact

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public String functionName() {
```
I see having "MatchFunction" and "QueryStringFunction" more explicit than "Match" and "Qstr" - happy to change that if you think it will make life harder for developers.

fang-xing-esql replied:
Could you please change the name of `MatchFunction` and `QueryStringFunction` to `Match` and `QueryString` respectively?

fang-xing-esql replied:
> Could you please change the name of `MatchFunction` and `QueryStringFunction` to `Match` and `QueryString` respectively?

Don't forget the change the name of these two classes.

Commit:
Move up method

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public String getWriteableName() {
```
Sure! 1c500ddaa309b8c49b96fa92cff073614af2a8bd

Commit:
Move up method

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:
```diff
@@ -84,4 +86,10 @@ protected NodeInfo<? extends Expression> info() {
     public String getWriteableName() {
```
They got together in dcc8bf6ed7e3f444c91427a43199e9b45bc10fb5

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,120 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+    public void testMatchFunctionNotAllowedWithNonPushableExpressions() {
+        assumeTrue("skipping because MATCH is not enabled", EsqlCapabilities.Cap.MATCH_FUNCTION.isEnabled());
+
+        assertEquals(
```
Agreed. I'll try to give this a go later - maybe in a follow up PR?

carlosdelest replied:
Conjunction and pushable disjunctions are supported now

Commit:
Remove unnecessary variable

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,120 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+    }
+
+    private void checkFullTextFunctionsOnlyAllowedInWhere(String functionName, String functionInvocation) throws Exception {
+        String function = functionName.toLowerCase(Locale.ROOT);
```
Good catch, fixed in 8b646a8490a262db607c7a73c2030c5f15c55a8d

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,120 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
         // Other value types are tested in QueryStringFunctionTests
     }
 
+    public void testMatchFunctionNullArgs() throws Exception {
```
Some of them are, but I thought it would get more confusing to add a common method when only 2 out of 5 tests are shared.

carlosdelest replied:
I'm sorry but the line numbers you mention are not referring specific queries. Can you link to the actual GH file + line number, or paste the query / context in the comment so I can check?

carlosdelest replied:
I see, thanks for the catch! Removed the duplicate tests in 419b8fbe6bb36b1dc4b3491be57631a44d3f2b05

carlosdelest commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -63,6 +63,18 @@ public static TypeResolution isString(Expression e, String operationName, ParamO
         return isType(e, DataType::isString, operationName, paramOrd, "string");
     }
 
+    public static TypeResolution isField(Expression e, String operationName, ParamOrdinal paramOrd) {
```
I see. I have implemented Validatable in 5a8a4f508e2403b9513a9813ad86d8b8c71eaf1f

carlosdelest replied:
One side effect of this, is that now checking for the column to be an actual ES field is not done until the `LogicalPlanOptimizer` kicks in.

The following does not trip a Verifier error:

`from test | sort emp_no | limit 10000 | where match(first_name, \"Anna\")"`

I'm ok with adding tests to the `LogicalPlanOptimizer` to catch this - just wanted to confirm that this is expected from the Verifier point of view.

carlosdelest commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -102,7 +102,13 @@ primaryExpression
+    : functionIdentifier LP (ASTERISK | (booleanExpression (COMMA booleanExpression)*))? RP
+    ;
+
+functionIdentifier
```
Did that in 79cd897df7e7ca257ab74c2f72a12213b020fb52

carlosdelest commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -102,7 +102,13 @@ primaryExpression
+functionIdentifier
+    : identifier
+    // Additional function identifiers that are already a reserved word in the language
+    | DEV_MATCH
```
Good point, added this in 79cd897df7e7ca257ab74c2f72a12213b020fb52

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
```
I did that in e44497800da5e4d7db87e80f765c22d5b99c39f4

carlosdelest replied:
I reverted back this change. I needed to include in serialization the query and the children, something that was not done in the original `QueryStringFunction` and caused BWC problems with previous versions of the function. 

We could introduce a TransportVersion to fix that, but given that there are different number of parameters that need to be serialized, having it done in the parent subclass was more confusing as it made the different `StreamInput` constructors more confusing.

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/FullTextFunction.java:
```diff
@@ -38,43 +37,80 @@ public static List<NamedWriteableRegistry.Entry> getNamedWriteables() {
+        return resolveNonQueryParamTypes().and(resolveQueryType());
+    }
+
+    private TypeResolution resolveQueryType() {
```
I changed the name and added some javadoc in 70ad4b0f8d98f44b01e482b8535d146424a95351 - hopefully it's more clear now!

Commit:
Fix deserialization

Commit:
Reimplement conditions to check for disjunctions

Commit:
Add tests for new checks

Commit:
Refactor checks using new Node method

Commit:
Adding disjunction tests

Commit:
Spotless

carlosdelest commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/tree/Node.java:
```diff
@@ -94,6 +95,27 @@ public <E extends T> void forEachUp(Class<E> typeToken, Consumer<? super E> acti
+     * @param action the action to execute for each parent of the specified typeToken
+     */
+    @SuppressWarnings("unchecked")
+    public <E extends T> E forEachParent(Class<E> typeToken, BiConsumer<E, ? super T> action) {
```
Added this method to help determining if a full text function has an unsupported parent (like IN or IS NULL). 

It seemed to me it could be useful in the future as it's a tree traversal method. LMK if you think there's a better alternative for this, or I should confine this to the Verifier class.

fang-xing-esql replied:
Could this be moved the `Verifier`? Just want to be cautious, `Node` is an important class, we change it only if it is necessary.

fang-xing-esql replied:
> Could this be moved the `Verifier`? Just want to be cautious, `Node` is an important class, we change it only if it is necessary.

Don't forget to move this into `Verifier`, and keep `Node` unchanged.

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +650,81 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
+        condition.forEachUp(Or.class, or -> {
+            Expression left = or.left();
+            Expression right = or.right();
+            left.forEachDown(FullTextFunction.class, ftf -> {
```
In case we have a full text function as a child of an OR expression, then the other side of it must be pushable to source.

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

Commit:
Reverted serialization change

fang-xing-esql reviewed (commented):
Thank you for lifting some of the restrictions related to `Match` and `QueryString`, it opens up a lot more opportunities for full text search in `ES|QL`. I think @nik9000 has good ideas for the potential/future usage of `Match` function, I added some more comments inspired by it, and I can imagine that it will open a lot more opportunities for `Match`, perhaps we can have a followup for short and long term.

Commit:
Add multivalued fields tests

Commit:
Refactor check into separate method

Commit:
Remove duplicate tests

Commit:
Add unsupported behaviour tests, and add limit checks for match function

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +650,81 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
+        });
+    }
+
+    private static void checkFullTextFunctionsParents(Expression condition, Set<Failure> failures) {
```
This check was added so the following is not allowed - would otherwise cause a downstream error:

`from test | where match(first_name, "Anna") in ("hello", "world")`

`from test | where match(first_name, "Anna") is null`

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,173 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+        // Other value types are tested in QueryStringFunctionTests
+    }
+
+    public void testQueryStringWithDisjunctionsThatCannotBePushedDown() {
```
Good catch, removed in 5f313999286e9fcf48ff580c143af8341c7b5421

carlosdelest commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1148,173 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+        checkWithDisjunctionsThatCannotBePushedDown("QSTR", "qstr(\"first_name: Anna\")");
+    }
+
+    public void testMatchWithDisjunctionsThatCannotBePushedDown() {
```
Yes! Removed in 5f313999286e9fcf48ff580c143af8341c7b5421

Commit:
Add mixed query string / match function CSV test

carlosdelest commented:
@fang-xing-esql :

> Can we cover some tests like below, even though they might not pass Verifier today, at least to capture the current behavior, these might become useful scenarios as Match function gets more popular.

Did that on 9dac2c69efb10ce24e0634f794c4dfec42617ccf

> This query throws VerificationException, gender is a Text field, I wonder if we expect it to work?

It works for me now that the OR restriction is lifted for pushable expressions - can you please check again?

> Another scenario that I can think of is mixed QSTR and Match in Where clause

That should be granted given that both are pushable expressions - there are multiple tests for that. But I added one in 3aa66af776216351a8155e0dc368aae934dc626c

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

ioanatia commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1087,6 +1088,13 @@ public void testMatchFilter() throws Exception {
     }
 
+    public void testMatchFunctionNotAllowedAfterCommands() throws Exception {
+        assertEquals(
```
do we want to add a `assumeTrue("skipping because MATCH is not enabled", EsqlCapabilities.Cap.MATCH_FUNCTION.isEnabled());` here?

carlosdelest replied:
Thanks for the catch! Done in 141dddab6a29967837c04d08a403f389f99c4e28

ChrisHegarty reviewed (approved):
LGTM

Commit:
Add assumeTrue check in test

fang-xing-esql commented:
> > This query throws VerificationException, gender is a Text field, I wonder if we expect it to work?
> 
> It works for me now that the OR restriction is lifted for pushable expressions - can you please check again?

It still fails for me. And it seems to be related to the use of `canPushToSource` in `Verifier`
```
from test
| where match(last_name, "Smith") or gender == "F"

Found 1 problem
line 2:9: Invalid condition [match(last_name, "Smith") or gender == "F"]. Function MATCH can't be used as part of an or condition that includes [gender == "F"]
org.elasticsearch.xpack.esql.VerificationException: Found 1 problem
```

fang-xing-esql reviewed (changes requested):
I'd suggest to evaluate whether using `canPushToSource` in `Verifier` is what we wanted, especially for `Text` fields related predicates used together with Match function in the same `Where` clause.

Commit:
Additional test for keyword fields being pushable to Lucene

carlosdelest commented:
> > > This query throws VerificationException, gender is a Text field, I wonder if we expect it to work?
> > 
> > 
> > It works for me now that the OR restriction is lifted for pushable expressions - can you please check again?
> 
> It still fails for me. And it seems to be related to the use of `canPushToSource` in `Verifier`

@fang-xing-esql  oh I see - in the employees dataset I have, `gender` is of type `keyword`, so I failed to reproduce this. Using a `keyword` field works as expected and it's possible to push down to Lucene - I've added this as a test in a25db902ffa1b23a80c712e078368d6c807304f1

I've checked with a text field, and the problem is that the text field is not considered as being able to be pushed down to Lucene - see [here](https://github.com/elastic/elasticsearch/blob/2f08d7d8be38e6b22624b192b3444730a2c62b54/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushDownUtils.java#L33).

We can revisit that code for text fields, but I believe it would be outside the scope of this PR.

Commit:
Removed duplicate tests

Commit:
Remove unnecessary concatenation

fang-xing-esql commented:
> > > > This query throws VerificationException, gender is a Text field, I wonder if we expect it to work?
> > > 
> > > 
> > > It works for me now that the OR restriction is lifted for pushable expressions - can you please check again?
> > 
> > 
> > It still fails for me. And it seems to be related to the use of `canPushToSource` in `Verifier`
> 
> @fang-xing-esql oh I see - in the employees dataset I have, `gender` is of type `keyword`, so I failed to reproduce this. Using a `keyword` field works as expected and it's possible to push down to Lucene - I've added this as a test in [a25db90](https://github.com/elastic/elasticsearch/commit/a25db902ffa1b23a80c712e078368d6c807304f1)
> 
> I've checked with a text field, and the problem is that the text field is not considered as being able to be pushed down to Lucene - see [here](https://github.com/elastic/elasticsearch/blob/2f08d7d8be38e6b22624b192b3444730a2c62b54/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushDownUtils.java#L33).
> 
> We can revisit that code for text fields, but I believe it would be outside the scope of this PR.

`Text` field does look a little tricky here, especially in terms of pushdown. I did some experiments, it is not always true that a `Text` field can or cannot be pushed down, it can be related to its subfield. 

```
curl -u elastic:password -X PUT "localhost:9200/books?pretty" -H 'Content-Type: application/json' -d'
{
  "mappings": {
    "properties": {
      "field1": {                      ===> predicate on field1 is not pushed down
        "type": "text"
      },
      "field2": {                     ===> predicate on field2 is pushed down
        "type": "text",
        "fields": {
           "keyword": {
            "type": "keyword"
           }
        }
      }
    }
  }
}
'

ProjectExec[[field1{f}#34, field2{f}#35, field2.keyword{f}#36]]                 = ProjectExec[[field1{f}#34, field2{f}#35, field2.keyword{f}#36]]
\_LimitExec[1000[INTEGER]]                                                      ! \_FieldExtractExec[field2{f}#35, field2.keyword{f}#36]<[]>
  \_FilterExec[field1{f}#34 == [61][KEYWORD] AND field2{f}#35 == [62][KEYWORD]] !   \_LimitExec[1000[INTEGER]]
    \_EsSourceExec[books][field1{f}#34, field2{f}#35, field2.keyword{f}#36]     !     \_FilterExec[field1{f}#34 == [61][KEYWORD]]
                                                                                !       \_FieldExtractExec[field1{f}#34]<[]>
                                                                                !         \_EsQueryExec[books], indexMode[standard], query[{"esql_single_value":{"field":"field2.keyword","next":{"term":{"field2.keyword":{"value":"b"}}},"source":"field2 == \"b\"@1:38"}}][_doc{f}#37], limit[], sort[] estimatedRowSize[null]
```

`hasIdenticalDelegate` is an input to `canPushToSource`, simply set it to false in `Verifier` may block some pushdown cases. It seems like the best time to decide whether a filter can be pushed down is at data nodes.

carlosdelest commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into carlosdelest/esql-match-function

Commit:
Move disjunctions check from Verifier to the LocalPhysicalVerifier

Commit:
Remove unneeded test

Commit:
Add disjunction that is pushable

Commit:
Spotless

Commit:
Merge remote-tracking branch 'carlosdelest/carlosdelest/esql-match-function' into carlosdelest/esql-match-function

carlosdelest commented:
@fang-xing-esql this PR keeps getting interesting 😅 

So in order to check that case, we need to perform the pushdown checks in the data nodes. 

I've done the following changes in c433aa1e657097d7329365cc44be4570981cbab7:
- Added a new dedicated `LocalPhysicalVerifier` class to perform the LocalPhysical plan verification. 
- This class receives the `LocalPhysicalOptimizerContext` in the `verify()` method so the `SearchStats` are available for verification
- I've moved the disjunction checks done in the `Verifier` to the `LocalPhysicalVerifier` class, and the corresponding test methods from `VerifierTests` to `LocalPhysicalVerifierTests`.
- I've written an additional check for the new behaviour in deb539b9b02c766eacf341ec8c28b9472f43d13f

Does that look good?

carlosdelest commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushDownUtils.java:
```diff
@@ -14,7 +14,7 @@
 import java.util.function.Predicate;
 
-class LucenePushDownUtils {
+public class LucenePushDownUtils {
```
Removed package visibility as this is now used from `LocalPhysicalVerifier`

carlosdelest commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into carlosdelest/esql-match-function

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,132 @@
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        // Query is the first child, field is the second child
+        return new MatchFunction(source(), newChildren.get(1), newChildren.getFirst());
```
We don't use `getFirst`, the majority of the functions call `newChildren.get(position)`. This is for the ESQL code to be consistent as much as possible in its entirety and not have multiple styles of writing code.

carlosdelest replied:
OK, done in 2ddcbb597956e1597964ba9b7003d581f73881f2

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,132 @@
+            description = "Text you wish to find in the provided field."
+        ) Expression matchQuery
+    ) {
+        super(source, matchQuery, List.of(matchQuery, field));
```
The order the children is inverted in `List.of(matchQuery, field)`. When `Function.nodeString()` is called (as part of `Node.toString()` which is essential when logging the tree) the match function will appear as `MATCH("abc", first_name)` (reversed as how the user is using it) which is confusing while looking at logs.

carlosdelest replied:
I see - done in 2ddcbb597956e1597964ba9b7003d581f73881f2

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:
```diff
@@ -61,27 +73,18 @@ public String functionName() {
     @Override
     public Expression replaceChildren(List<Expression> newChildren) {
-        return new QueryStringFunction(source(), newChildren.get(0));
+        return new QueryStringFunction(source(), newChildren.getFirst());
```
Same as `Match` function, use `newChildren.get(position)` instead of `getFirst()`

carlosdelest replied:
OK, done in 2ddcbb597956e1597964ba9b7003d581f73881f2

fang-xing-esql commented on x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4:
```diff
@@ -102,7 +102,13 @@ primaryExpression
+
+functionName
+    // Additional function identifiers that are already a reserved word in the language
+    : {this.isDevVersion()}? DEV_MATCH
```
The `Match` command is gone, it is a good time to remove the `DEV_MATCH` command from lexer and consolidate it with the `DEV_MATCH_OP`.

carlosdelest replied:
Makes sense, done in 19f05465e146f831807aa0f714074ac5192fce6e

Commit:
Replace getFirst() for get(0)

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchFunction.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    public Query asQuery(String queryText) {
+        return new MatchQuery(source(), ((FieldAttribute) field).name(), queryText);
```
Move the build of `MatchQuery` into `EsqlExpressionTranslators`, it is a central place to build pushdown queries.

carlosdelest replied:
That allows to simplify the methods in the superclass, done in 06f2eafd9177a8d2d8c80300072039ac9e79407e

Commit:
Replace getFirst() for get(0), change children list order

Commit:
Regenerated from tests

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:
```diff
@@ -61,27 +73,18 @@ public String functionName() {
-            throw new IllegalArgumentException("Query in QSTR needs to be resolved to a string");
-        }
+    public Query asQuery(String queryText) {
+        return new QueryStringQuery(source(), queryText, Map.of(), null);
```
Move the build of QueryStringQuery into EsqlExpressionTranslators, it is a central place to build pushdown queries.

carlosdelest replied:
Done in 06f2eafd9177a8d2d8c80300072039ac9e79407e

Commit:
Removed MATCH command from the lexer

fang-xing-esql commented:
> @fang-xing-esql this PR keeps getting interesting 😅
> 
> So in order to check that case, we need to perform the pushdown checks in the data nodes.
> 
> I've done the following changes in [c433aa1](https://github.com/elastic/elasticsearch/commit/c433aa1e657097d7329365cc44be4570981cbab7):
> 
> * Added a new dedicated `LocalPhysicalVerifier` class to perform the LocalPhysical plan verification.
> * This class receives the `LocalPhysicalOptimizerContext` in the `verify()` method so the `SearchStats` are available for verification
> * I've moved the disjunction checks done in the `Verifier` to the `LocalPhysicalVerifier` class, and the corresponding test methods from `VerifierTests` to `LocalPhysicalVerifierTests`.
> * I've written an additional check for the new behaviour in [deb539b](https://github.com/elastic/elasticsearch/commit/deb539b9b02c766eacf341ec8c28b9472f43d13f)
> 
> Does that look good?

Thanks for exploring @carlosdelest! Indeed it is getting more interesting when we look at it closer. Let's hold off `LocalPhysicalVerifier`, we know that calling `canPushToSource` from the coordinator is not quite right either. We have explored and knew the limitations with pushing down `Match` function with disjunctions. Lifting the limitations on conjunctions is a good step forward. I'd suggest to limit the scope of this PR to allow `Match` and `QueryString` with conjunctions only and create a separate issue to tackle `Match` and `QueryString` functions with disjunctions.

Commit:
Moved query generation to EsqlExpressionTranslators, simplified class hierarchy methods

Commit:
Revert "Move disjunctions check from Verifier to the LocalPhysicalVerifier"

This reverts commit c433aa1e

Commit:
Revert "Add disjunction that is pushable"

This reverts commit deb539b9b02c766eacf341ec8c28b9472f43d13f.

Commit:
Revert "Remove unneeded test"

This reverts commit 2200f4d2a9c12bcbb3e2363006a4c57b57a57a76.

Commit:
Remove LocalPhysicalVerifier

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:
```diff
@@ -1147,26 +1156,115 @@ public void testQueryStringFunctionsNotAllowedAfterCommands() throws Exception {
+            "1:19: second argument of [match(first_name, null)] cannot be null, received [null]",
+            error("from test | where match(first_name, null)")
+        );
+    }
```
This query doesn't complain about the null input, and return no result. Can match return null?
`row n = null | eval text = n + 5 | where match(text::keyword, "abc")`

carlosdelest replied:
I checked - this is because the full text functions were nullable and this was optimized and thus removing the actual function. I fixed that by marking the functions as not nullable, and adding a test in 937ddf1c25aa252d25402b890791c8fcc0b12756

Commit:
Merge remote-tracking branch 'carlosdelest/carlosdelest/esql-match-function' into carlosdelest/esql-match-function

fang-xing-esql commented:
This is a snapshot function, it will be great to add `test-release` label to catch possible surprises earlier.

carlosdelest commented:
> Let's hold off LocalPhysicalVerifier

😓 reverted. It was fun.

carlosdelest commented:
> This is a snapshot function, it will be great to add test-release label to catch possible surprises earlier.

Makes sense, but adding that will add time to the CI checks. Let's do that once the PR is approved.

Commit:
Fix usage after ROW for match function

Commit:
Reverted row change

Commit:
FullTextFunctions are not nullable

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

# Conflicts:
#	x-pack/plugin/esql/src/main/antlr/EsqlBaseLexer.tokens
#	x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.tokens
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseLexer.interp
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseLexer.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser.interp
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser.java

Commit:
Regenerate lexer / parser with changes from main

Commit:
Additional assumeTrue that were missing

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:
```diff
@@ -645,27 +651,91 @@ private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> fa
+
+    private static void checkDisjunction(Set<Failure> failures, Or or, Expression left, Expression right) {
+        left.forEachDown(FullTextFunction.class, ftf -> {
+            if (canPushToSource(right, x -> false) == false) {
```
Calling `canPushToSource` from the coordinator is not quite right. I'd suggests to disable `Match` with disjunctions for now as a known limitation, and open a separate issue to find a solution in general.

carlosdelest replied:
OK, let's disable this entirely and use a follow up issue. Done in 0dd782c2799b4d5f3c6c9e2d0d3d771dc298bf45

fang-xing-esql commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:
```diff
@@ -571,11 +737,126 @@ public void testQueryStringFunctionMultipleWhereClauses() {
+                    "1:19: Invalid condition [({} and emp_no < 1000) or (length(first_name) > 12 and emp_no > 1000)]. "
+                        + "Function {} can't be used as part of an or condition that includes [length(first_name) > 12 and emp_no > 1000]",
+                    functionInvocation,
+                    functionName
+                )
+            )
+        );
+    }
+
+    public void testMatchFunctionDisjunctionNonPushableClauses() {
+        assumeTrue("skipping because MATCH function is not enabled", EsqlCapabilities.Cap.MATCH_FUNCTION.isEnabled());
+        String queryText = """
+            from test
+            | where match(first_name, "Peter") or length(last_name) > 10
+            """;
+
+        VerificationException ve = expectThrows(VerificationException.class, () -> plannerOptimizer.plan(queryText, IS_SV_STATS));
+        assertThat(
+            ve.getMessage(),
+            containsString(
+                "Invalid condition [match(first_name, \"Peter\") or length(last_name) > 10]. "
+                    + "Function MATCH can't be used as part of an or condition that includes [length(last_name) > 10]"
+            )
+        );
+    }
+
+    public void testMatchFunctionIsNotNullable() {
+        assumeTrue("skipping because MATCH function is not enabled", EsqlCapabilities.Cap.MATCH_FUNCTION.isEnabled());
+
+        String queryText = """
+            row n = null | eval text = n + 5 | where match(text::keyword, "Anna")
+            """;
+
+        VerificationException ve = expectThrows(VerificationException.class, () -> plannerOptimizer.plan(queryText, IS_SV_STATS));
+        assertThat(
+            ve.getMessage(),
+            containsString("[MATCH] cannot operate on [text::keyword], which is not a field from an index mapping")
+        );
+    }
+
```
Move these tests catching `VerificationException` to `LogicalPlanOptimizerTests`.

carlosdelest replied:
Done in 87807ed508002848138e2871620fe2a7c6ce61fc

fang-xing-esql reviewed (commented):
Thank you for addressing some of the comments from yesterday @carlosdelest . Some of them that you might have missed are highlighted below.

Commit:
Full text functions can't be used as part of OR conditions

Commit:
Change QueryStringFunction and MatchFunction to QueryString and Match

Commit:
Moved forEachParent method from Node to Verifier

Commit:
Moved verification tests to LogicalPlanOptimizerTests

Commit:
Simplified checking of full text function parents

carlosdelest commented:
Sorry @fang-xing-esql , but I didn't see any additional comments nor can I find them now. Thanks for pasting them again.

> Don't forget the change the name of these two classes.

Done in d4e625696529e6b81231b1bbc3691b095344b6b0

> Don't forget to move this into Verifier, and keep Node unchanged.

OK, moved in 03a02460b9934204812c91d3da300afb7c19fc13 and simplified in 2e5ce4ede88f199eab5d218646c43f7db4436e26

carlosdelest commented:
@fang-xing-esql , I think I have addressed all comments as of now. 

If there are no more comments to address, let me know and I'll add the `test-release` label for a final check.

carlosdelest commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into carlosdelest/esql-match-function

carlosdelest commented:
@fang-xing-esql , I added the `test-release` label and the tests are green. Are there any other remaining concern for this function to be merged?

Commit:
Merge branch 'main' into carlosdelest/esql-match-function

# Conflicts:
#	x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseLexer.interp
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseLexer.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser.interp
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlBaseParser.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/EsqlParser.java
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java

Commit:
Fix merge

fang-xing-esql reviewed (approved):
LGTM! Thanks a lot for all the efforts and being patient with me @carlosdelest ! 

I wonder if there is an issue created for keeping track of full text functions with disjunctions(OR) yet? Can we have it linked to this PR, so that we don't forget about it? Thanks again!

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 113374`

carlosdelest commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/114695">](https://github.com/elastic/elasticsearch/pull/114695)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.5.1 -->
